### PR TITLE
Video daemon

### PR DESCRIPTION
**Pull request by griffobeid**
## How to test:

```
RUST_LOG=info cargo run --bin video-daemon --release -- --meeting-id abc --user-id griffin1 https://transport.rustlemania.com
```

## Known Issues

- Decode errors on UI

### CODE REVIEW

**Review by darioalessandro — APPROVED** [review]
good for a POC! 

I will test it tonight buddy!!

File: video-daemon/src/main.rs [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+    tracing::subscriber::set_global_default(
+        tracing_subscriber::FmtSubscriber::builder()
+            .with_env_filter(tracing_subscriber::EnvFilter::from_default_env())
+            .finish(),
+    )
+    .unwrap();
+    let opt = Opt::parse();
+    let mut client = Client::new(opt).expect("failed to create client");
+    client.connect().await.expect("failed to connect");
+    let camera_config = CameraConfig {
```
Comment (darioalessandro) [inline video-daemon/src/main.rs:19]: This is starting to look better!!

File: video-daemon/src/main.rs [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+        height: 480,
+        framerate: 30,
+        frame_format: nokhwa::utils::FrameFormat::YUYV,
+        video_device_index: 0,
+    };
+    let (quic_tx, mut quic_rx) = channel::<Vec<u8>>(1);
+    let mut camera = CameraDaemon::from_config(camera_config, quic_tx);
+    camera.start().expect("failed to start camera");
+    while let Some(data) = quic_rx.recv().await {
+        if let Err(e) = client.send(data).await {
```
Comment (darioalessandro) [inline video-daemon/src/main.rs:30]: might as well make it crash or terminate

File: actix-api/src/webtransport/mod.rs [hunk 3]
```diff
@@ -109,27 +113,41 @@ pub async fn start(opt: WebTransportOpt) -> Result<(), Box<dyn std::error::Error
...
-
-                    // info!("Establishing WebTransport session");
-                    // // 3. TODO: Conditionally, if the client indicated that this is a webtransport session, we should accept it here, else use regular h3.
-                    // // if this is a webtransport session, then h3 needs to stop handing the datagrams, bidirectional streams, and unidirectional streams and give them
-                    // // to the webtransport session.
-                    let nc = nc.clone();
-                    if let Err(err) = handle_connection(h3_conn, nc).await {
-                        error!("Failed to handle connection: {err:?}");
+                    let mut http3 = false;
+                    if let Some(data) = conn.handshake_data() {
```
[the remark this answers is not in the document]
Comment (darioalessandro) [reply actix-api/src/webtransport/mod.rs:127]: @griffobeid ^^

File: yew-ui/src/components/attendants.rs [hunk 4]
```diff
@@ -255,7 +255,7 @@ impl Component for AttendantsComponent {
                 { self.error.as_ref().map(|error| html! { <p>{ error }</p> }) }
                 { rows }
                 {
-                    if USERS_ALLOWED_TO_STREAM.iter().any(|host| host == &email) || USERS_ALLOWED_TO_STREAM.is_empty() {
+                    if USERS_ALLOWED_TO_STREAM.len() == 0 || USERS_ALLOWED_TO_STREAM.iter().any(|host| host == &email) {
```
Comment (darioalessandro) [inline yew-ui/src/components/attendants.rs:258]: This change looks unrelated, keep in mind that I had to fix this in master: https://github.com/security-union/videocall-rs/pull/145

File: videocall-client/src/media_devices/media_device_list.rs [hunk 5]
```diff
@@ -144,8 +144,8 @@ impl MediaDeviceList {
         let on_loaded = self.on_loaded.clone();
         let on_audio_selected = self.audio_inputs.on_selected.clone();
         let on_video_selected = self.video_inputs.on_selected.clone();
-        let audio_input_devices = Arc::clone(&self.audio_inputs.devices);
-        let video_input_devices = Arc::clone(&self.video_inputs.devices);
+        let audio_input_devices = Rc::clone(&self.audio_inputs.devices);
```
Comment (darioalessandro) [inline videocall-client/src/media_devices/media_device_list.rs:147]: @griffobeid did clippy do this?

Comment (griffobeid) [reply videocall-client/src/media_devices/media_device_list.rs:147]: Yeah https://rust-lang.github.io/rust-clippy/master/index.html#/Arc

File: Dockerfile.video-daemon [hunk 6]
```diff
@@ -0,0 +1,38 @@
+FROM debian:bookworm-slim as build
```
Comment (griffobeid) [inline Dockerfile.video-daemon:1]: This should use the api base image

File: actix-api/src/webtransport/mod.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (griffobeid) [reply actix-api/src/webtransport/mod.rs]: Moved this logic out to a function, when I tried to write it with the chained and_then I was getting a lot of ownership issues

File: video-daemon/src/main.rs (hunk 2 above)
Comment (griffobeid) [reply video-daemon/src/main.rs]: done

File: videocall-client/src/media_devices/media_device_list.rs (hunk 5 above)
Comment (darioalessandro) [reply videocall-client/src/media_devices/media_device_list.rs:147]: neat!

**Comment by darioalessandro** [comment]
@griffobeid I got it to work :) 

Please take a look at the reconnect logic because right now is just making the video-daemon crash.

https://github.com/security-union/videocall-rs/assets/1176339/2e2fa2ef-50ec-47f3-bff1-4cb7af4aab6e

**Pull request merged**